[Merged by Bors] - Fix color banding by dithering image before quantization

aevyrie opened the pull request:
# Objective

- Closes #5262 
- Fix color banding caused by quantization.

## Solution

- Adds dithering to the tonemapping node from #3425.
- This is inspired by Godot's default "debanding" shader: https://gist.github.com/belzecue/c4b0bdede4e6d4b948679011096be150
- Unlike Godot:
  - debanding happens after tonemapping. My understanding is that this is preferred, because we are running the debanding at the last moment before quantization (`[f32, f32, f32, f32]` -> `f32`). This ensures we aren't biasing the dithering strength by applying it in a different (linear) color space.
  - This code instead uses and reference the origin source, Valve at GDC 2015

![Screenshot from 2022-11-10 13-44-46](https://user-images.githubusercontent.com/2632925/201218880-70f4cdab-a1ed-44de-a88c-8759e77197f1.png)
![Screenshot from 2022-11-10 13-41-11](https://user-images.githubusercontent.com/2632925/201218883-72393352-b162-41da-88bb-6e54a1e26853.png)


## Additional Notes 

Real time rendering to standard dynamic range outputs is limited to 8 bits of depth per color channel. Internally we keep everything in full 32-bit precision (`vec4<f32>`) inside passes and 16-bit between passes until the image is ready to be displayed, at which point the GPU implicitly converts our `vec4<f32>` into a single 32bit value per pixel, with each channel (rgba) getting 8 of those 32 bits.

### The Problem

8 bits of color depth is simply not enough precision to make each step invisible - we only have 256 values per channel! Human vision can perceive steps in luma to about 14 bits of precision. When drawing a very slight gradient, the transition between steps become visible because with a gradient, neighboring pixels will all jump to the next "step" of precision at the same time.

### The Solution

One solution is to simply output in HDR - more bits of color data means the transition between bands will become smaller. However, not everyone has hardware that supports 10+ bit color depth. Additionally, 10 bit color doesn't even fully solve the issue, banding will result in coherent bands on shallow gradients, but the steps will be harder to perceive.

The solution in this PR adds noise to the signal before it is "quantized" or resampled from 32 to 8 bits. Done naively, it's easy to add unneeded noise to the image. To ensure dithering is correct and absolutely minimal, noise is adding *within* one step of the output color depth. When converting from the 32bit to 8bit signal, the value is rounded to the nearest 8 bit value (0 - 255). Banding occurs around the transition from one value to the next, let's say from 50-51. Dithering will never add more than +/-0.5 bits of noise, so the pixels near this transition might round to 50 instead of 51 but will never round more than one step. This means that the output image won't have excess variance:
  - in a gradient from 49 to 51, there will be a step between each band at 49, 50, and 51.
  - Done correctly, the modified image of this gradient will never have a adjacent pixels more than one step (0-255) from each other.
  - I.e. when scanning across the gradient you should expect to see:
```
                  |-band-| |-band-| |-band-|
Baseline:         49 49 49 50 50 50 51 51 51
Dithered:         49 50 49 50 50 51 50 51 51
Dithered (wrong): 49 50 51 49 50 51 49 51 50
```

![Screenshot from 2022-11-10 14-12-36](https://user-images.githubusercontent.com/2632925/201219075-ab3f46be-d4e9-4869-b66b-a92e1706f49e.png)
![Screenshot from 2022-11-10 14-11-48](https://user-images.githubusercontent.com/2632925/201219079-ec5d2add-817d-487a-8fc1-84569c9cda73.png)




You can see from above how correct dithering "fuzzes" the transition between bands to reduce distinct steps in color, without adding excess noise.

### HDR

The previous section (and this PR) assumes the final output is to an 8-bit texture, however this is not always the case. When Bevy adds HDR support, the dithering code will need to take the per-channel depth into account instead of assuming it to be 0-255. Edit: I talked with Rob about this and it seems like the current solution is okay. We may need to revisit once we have actual HDR final image output.

---

## Changelog

### Added

- All pipelines now support deband dithering. This is enabled by default in 3D, and can be toggled in the `Tonemapping` component in camera bundles. Banding is a graphical artifact created when the rendered image is crunched from high precision (f32 per color channel) down to the final output (u8 per channel in SDR). This results in subtle gradients becoming blocky due to the reduced color precision. Deband dithering applies a small amount of noise to the signal before it is "crunched", which breaks up the hard edges of blocks (bands) of color. Note that this does not add excess noise to the image, as the amount of noise is less than a single step of a color channel - just enough to break up the transition between color blocks in a gradient.

alice-i-cecile commented:
Please note that this fixes #5262 :)

aevyrie commented:
FYI I'm planning on making an example to better stress test noise and banding performance before removing draft status.

bjorn3 commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -98,8 +106,10 @@ fn fragment(in: FragmentInput) -> [[location(0)]] vec4<f32> {
 
     return output_color;
-}
+}
```
nit: missing trailing newline

bjorn3 commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -24,6 +24,14 @@ struct FragmentInput {
+    dither = fract(dither.rgb / vec3<f32>(103.0, 71.0, 97.0));
+	// Subtract 0.5 to avoid slightly brightening the whole viewport.
+    return vec4<f32>((dither.rgb - 0.5) / 255.0, 0.0);
+}
```
Where did this code originate from? Did you modify it for VR yourself, or did someone else do so?

aevyrie replied:
Another reason this is marked as draft - I haven't looked at the licenses for any of my references. Might need some support for that. Unsure if little code snippets like this are acceptable.

aevyrie commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -98,8 +106,10 @@ fn fragment(in: FragmentInput) -> [[location(0)]] vec4<f32> {
-        output_color = tone_mapping(pbr(pbr_input));
+        let pbr_out = pbr(pbr_input);
+        output_color = screen_space_dither(in.frag_coord.xy) * min(15.0 * pbr_out, vec4<f32>(1.0)) + pbr_out;
+        output_color = tone_mapping(output_color);
```
Thanks for the heads up. I'll probably just rebase on that and add a debanding node, or put it in tonemap.

aevyrie commented:
Added a stress test for banding:
### Control
![image](https://user-images.githubusercontent.com/2632925/178193518-78e19e88-9aa5-4074-8caa-16311543a6c3.png)
### Dithered
![image](https://user-images.githubusercontent.com/2632925/178193405-98086ab3-c6d8-4355-a390-0b2c6ecd75d7.png)

aevyrie commented:
@CptPotato I rebased on #3425, and added it to the tonemapping pass. This means dithering is now a fullscreen post-process. Could probably improve by moving to its own node that outputs to a `Bgra8Unorm` (instead of SRGB, I think this is correct?) to remove the unnecessary conversion to and from SRGB space.

devil-ira commented:
As a note to others viewing the images above if you still see banding in the dithered image click on it to view it in another tab.
That will display the image at 100% scaling and the banding should not be visible.

aevyrie commented:
> The result looks good! If tonemapping is disabled, will dithering be disabled too?

Yup, that's correct. Maybe that make sense though? The goal of dithering is to remove artifacts from quantization, which seems to fit alongside tonemapping (mapping between color spaces).

> [...] you could probably eliminate the conversion back to linear by changing the format of the render target view ("attachment" I think is the wgpu slang).

The issue with this is device compatibility from what @superdump mentioned on discord.

> Though, in that case a more accurate conversion function than `pow(col, 1.0/2.2)` would make sense.

Any resources on what that would be?

aevyrie commented:
Thanks for the links!

> That being said, if you keep the conversion in both directions (linear -> srgb -> linear) in the shader the difference it makes on the dithering may be negligible. I can't say for sure, though.

This is a good point. Dithering just relies on the rounding being correct, such that dithering never causes more than one full step away from the true signal value during quantization. As long as the log approximation is "pretty close" (a rigorous technical term), it wouldn't cause quantization to round another full step for the large majority of cases.

superdump commented:
> > [...] you could probably eliminate the conversion back to linear by changing the format of the render target view ("attachment" I think is the wgpu slang).
> 
> The issue with this is device compatibility from what @superdump mentioned on discord.

This is only relevant if rendering directly into the swapchain texture. The formats supported by the swapchain are more limited as they're specifically for outputting to a display. If rendering to an intermediate texture then the format support is much broader. The tonemapping / upscaling PR does allow for using an independent intermediate texture rather than rendering directly to the swapchain texture, so in some cases at least we could do this. Basically probably only the final swapchain texture might need to be an srgb format and we can use non-srgb in-between. We could even do gamma correction manually and not use an srgb format at all, though I'm not sure of exact swapchain texture format support across a variety of different backends / GPU vendors though so it may be that some/many only support srgb formats.

superdump commented:
I asked Alex Vlachos about the license of the GDC 2015 code: https://twitter.com/AlexVlachos/status/1557662473775239175

aevyrie commented:
@mockersf and @cart, RE: discord, I've cleaned this up and got it working with and without hdr.

superdump commented:
I've reloaded this information into my mental cache and had a think through the current state of things.

SDR:
* main pass f32 with tonemapping -> srgb u8 texture
* Dithering should happen in the main pass on non-linear sRGB values, then be converted back to linear just before writing out

'HDR' (it's still SDR at the end as we don't have HDR output to the window yet, but it allows us to do post-processing on linear HDR before tonemapping):
* main pass f32 -> linear f16 texture -> tonemapping f32 -> linear f16 texture -> upscaling f32 -> srgb u8 texture
* Dithering should happen in the upscaling pass on non-linear sRGB values, then be converted back to linear just before writing out

I think this should be correct...

superdump commented:
Nope, I forgot about upscaling in the SDR case:
* main pass f32 with tonemapping -> srgb u8 texture -> upscaling f32 -> srgb u8 texture
* Does dithering need to be done twice to compensate for both the intermediate quantisation errors and the upscaled errors? i.e. in both the main pass and upscaling pass?

cart commented:
This should be disable-able (probably on the existing Tonemapping settings, given that they are currently tied together). Disabling dither there should specialize the tonemapping pipeline to remove the dither code.

aevyrie commented:
I'm not super proficient with pipeline specialization, but I think I've completed this as requested. Dithering can be toggled, but requires tonemapping to be enabled. This is reflected in the new type signature of `Tonemapping`.

mockersf commented on crates/bevy_core_pipeline/src/tonemapping/tonemapping_shared.wgsl:
```diff
@@ -27,3 +27,11 @@ fn reinhard_luminance(color: vec3<f32>) -> vec3<f32> {
+    var dither = vec3<f32>(dot(vec2<f32>(171.0, 231.0), frag_coord)).xxx;
+    dither = fract(dither.rgb / vec3<f32>(103.0, 71.0, 97.0));
+    return (dither - 0.5) / 255.0;
+}
```
```suggestion
}

```

mockersf commented:
just checking, as it's not what I understood from the discussion: should this dithering setting be ignored when HDR is enabled?

cart commented:
> just checking, as it's not what I understood from the discussion: should this dithering setting be ignored when HDR is enabled?

Nope. Dithering should be coupled to tonemapping (at least for now, as they are logically coupled in the hdr tonemapping pass). If HDR is enabled, dithering should happen in the tonemapping pass. If HDR is disabled, it should happen in the PBR / sprite shader.

This impl doesn't quite line up with that, as we aren't specializing the tonemapping pipeline when dither is enabled, which means it is skipped when hdr is enabled.

cart commented:
Working on a fix now

cart commented:
Fixed! I also made a personal style call by renaming `is_deband_dither_enabled` to `deband_dither`. I like preferring the shorter names when the word is already a verb. I think they read better by nature of having less grammar to mentally parse.

aevyrie commented:
> Fixed! I also made a personal style call by renaming `is_deband_dither_enabled` to `deband_dither`. I like preferring the shorter names for bools when the word is already a verb. I think they read better by nature of having less grammar to mentally parse.

Thanks. I agree with the rename, I had the same initially but chose to follow the preexisting `is_` style. :slightly_smiling_face:

superdump commented on crates/bevy_core_pipeline/src/tonemapping/tonemapping.wgsl:
```diff
@@ -10,5 +10,15 @@ var hdr_sampler: sampler;
     let hdr_color = textureSample(hdr_texture, hdr_sampler, in.uv);
 
-    return vec4<f32>(reinhard_luminance(hdr_color.rgb), hdr_color.a);
+    var output_rgb = vec3<f32>(reinhard_luminance(hdr_color.rgb));
```
Nit: if Reinhard luminance returns a vec3 then wrapping it in one is unnecessary.

aevyrie replied:
The intermediate variable is needed to work with the ifdef, and reassemble a vec4 with the alpha value. Right?

superdump replied:
I wasn’t suggesting to remove the variable. I was just saying that if the function returns a vec3 then we don’t need to do vec3(func()), just func().

aevyrie replied:
Oh, duh! Thanks for clarifying. 🙂

superdump reviewed (approved):
I’m marking this as approved assuming it is all tested in its current state as I only had a couple of nit comments.

cart commented:
bors r+
